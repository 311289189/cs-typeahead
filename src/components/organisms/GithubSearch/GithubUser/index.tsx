import React from 'react';
import classnames from 'classnames'

import { User } from "../../../../../types/github";
const styles = require('./styles.module.css')

interface Props {
  user: User
  isLoading: boolean
}

export default ({ user, isLoading }: Props) => (
  <a
    href={user.html_url}
    target="_blank"
    rel="noopener noreferrer"
    className={classnames(styles.userTile, { [styles.loading]: isLoading })}
  >
    <img
      className={styles.avatar}
      src={user.avatar_url}
      alt={user.login}
    />
    <div className={styles.userLogin}>{user.login}</div>
  </a>
)
