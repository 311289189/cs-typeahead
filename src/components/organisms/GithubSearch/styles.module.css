:root {
    --small-screen-offset: 20px;
}

.userTilesWrapper {
    position: absolute;
    background: white;
    overflow-y: scroll;
    border-radius: 0 0 10px 10px;
    height: 0;
    top: 30px;
    width: 100%;
    transition: height 200ms;
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
}

.userTilesWrapper.isOpen {
    height: 50vh;
}

.errorMessage {
    margin: 20px 0;
}

.searchInput {
    display: flex;
    position: relative;
    justify-content: center;
    flex-direction: column;
    width: 50vw;
    margin: auto;
}

@media (max-width: 600px) {
    .searchInput {
        width: auto;
        padding: 0 var(--small-screen-offset);
    }
    .userTilesWrapper.isOpen {
        width: calc(100% - var(--small-screen-offset) * 2);
        left: var(--small-screen-offset);
    }
}

.inputField {
    font-size: 15px;
    border: 5px solid var(--color-languidLavender);
    border-radius: 5px 5px 0 0;
    outline: none;
    color: var(--color-MsuGreen)
}

.hideScrollbar {
    scrollbar-width: none; /* Firefox 64 */
    -ms-overflow-style: none; /* Internet Explorer 11 */
}

.hideScrollbar::-webkit-scrollbar { /** WebKit */
    display: none;
}
