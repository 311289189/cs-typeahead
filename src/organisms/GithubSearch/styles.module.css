.userTilesWrapper {
    position: absolute;
    background: white;
    overflow-y: scroll;
    border-radius: 0 0 10px 10px;
    height: 0;
    top: 30px;
    width: 100%;
    transition: height 200ms;
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
}

.userTilesWrapper.isOpen {
    height: 50vh;
}

.errorMessage {
    margin: 20px 0;
}

.searchInput {
    display: flex;
    position: relative;
    justify-content: center;
    flex-direction: column;
    width: 50vw;
    margin: auto;
}

.inputField {
    font-size: 15px;
    border: 5px solid #DDC4DD;
    border-radius: 5px 5px 0 0;
    outline: none;
}

.hideScrollbar {
    scrollbar-width: none; /* Firefox 64 */
    -ms-overflow-style: none; /* Internet Explorer 11 */
}

.hideScrollbar::-webkit-scrollbar { /** WebKit */
    display: none;
}
