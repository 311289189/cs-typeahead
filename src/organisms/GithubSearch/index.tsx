import React from 'react';
import classnames from 'classnames';

import InputField from "../../components/InputField";
import useDebounce from "../../hooks/useDebounce";
import apiFetch from "../../utils/apiFetch";
import {GithubUsersResponse, User} from "../../../types/github";
import GithubUser from "./GithubUser";
import {HTTP_URL} from "../../config/http";
const styles = require('./styles.module.css');

const GithubSearch = () => {
  const [searchTerm, setSearchTerm] = React.useState<string>('');
  const [isSearching, setIsSearching] = React.useState<boolean>(false);
  const [results, setResults] = React.useState<User[]>([]);
  const [searchFocus, setSearchFocus] = React.useState<boolean>(false);
  const [error, setError] = React.useState<string>('');
  const debouncedSearchTerm = useDebounce<string>(searchTerm, 500);

  React.useEffect(() => {
    if (debouncedSearchTerm) {
      setError('');
      setIsSearching(true);

      apiFetch<GithubUsersResponse>(`${HTTP_URL.GITHUB}?q=${debouncedSearchTerm}`).then(({ json, data }) => {
        if (data.status !== 200) {
          setError(data.status === 403 ? '403 Forbidden: Please check your token' : 'API Error')
        } else {
          setResults(json.items);
        }
        setIsSearching(false);
      });
    } else {
      setResults([]);
    }
  // eslint-disable-next-line
  }, [debouncedSearchTerm])

  const isUserTyping = searchTerm !== debouncedSearchTerm

  return (
    <main>
      <div className={styles.searchInput}>
        <InputField
          className={styles.inputField}
          hasFocus={setSearchFocus}
          placeholder="Search for a user: "
          id="searchInput"
          onChange={setSearchTerm}
          value={searchTerm}
        />
        <div className={styles.errorMessage}>{ error && error }</div>
        <div className={classnames(styles.userTilesWrapper, styles.hideScrollbar, {
          [styles.isOpen]: results && results.length && searchTerm && searchFocus && !error
        })}>
          {results && results.map((result, index) => (
            <GithubUser
              isLoading={isSearching || isUserTyping}
              key={index}
              user={result}
            />
          ))}
        </div>
      </div>
    </main>
  );
}

export default GithubSearch
