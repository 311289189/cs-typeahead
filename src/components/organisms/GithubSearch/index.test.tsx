import React from 'react'
import { render } from '@testing-library/react';

import GithubSearch from "./index";

describe('<GithubSearch> Component', () => {
  it('renders', () => {
    const { baseElement } = render(<GithubSearch />);
    expect(baseElement).toBeInTheDocument();
  })
})
