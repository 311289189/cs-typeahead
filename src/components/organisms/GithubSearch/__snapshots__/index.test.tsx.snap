// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GithubSearch> Component renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div>
          <label
            for="apiKey"
          >
            If you would like to use your own API key, please enter it here 
          </label>
          <input
            class="generic-input"
            id="apiKey"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="searchInput"
          >
            Search for a user: 
          </label>
          <input
            class="generic-input"
            id="searchInput"
            type="text"
            value=""
          />
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <div>
        <label
          for="apiKey"
        >
          If you would like to use your own API key, please enter it here 
        </label>
        <input
          class="generic-input"
          id="apiKey"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="searchInput"
        >
          Search for a user: 
        </label>
        <input
          class="generic-input"
          id="searchInput"
          type="text"
          value=""
        />
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
